src/Pos/Delegation/Logic.hs

```diff
@@ -171,40 +133,294 @@ initDelegation = do
+    Nothing -> GS.getPskByIssuer $ Left iPk
+    Just s  -> pure $ Just s
+
+-- Copied from 'these' library.
+data These a b = This a | That b | These a b
+    deriving (Eq, Show, Generic)
+
+-- u → (Maybe d₁, Maybe d₂): u changed delegate from d₁ (or didn't
+-- have one) to d₂ (or revoked delegation). These a b ≃ (Maybe a,
+-- Maybe b) w/o (Nothing,Nothing).
+type TransChangeset = HashMap PublicKey (These PublicKey PublicKey)
+type ReverseTrans = HashMap PublicKey (HashSet PublicKey, HashSet PublicKey)
+
+-- WHENEVER YOU CHANGE THE FUNCTION, CHECK DOCUMENTATION CONSISTENCY! THANK YOU!
+--
+-- Takes a set of dlg edge actions to apply and returns compensations
+-- to dlgTrans and dlgTransRev parts of delegation db. Should be
+-- executed under shared Gstate DB lock.
+calculateTransCorrections
+    :: forall m.
+       MonadDBPure m
+    => HashSet GS.DlgEdgeAction -> m SomeBatchOp
+calculateTransCorrections eActions = do
+    -- Get the changeset and convert it to transitive ops.
+    changeset <- transChangeset
+    let toTransOp iPk (This _)      = GS.DelTransitiveDlg iPk
+        toTransOp iPk (That dPk)    = GS.AddTransitiveDlg iPk dPk
+        toTransOp iPk (These _ dPk) = GS.AddTransitiveDlg iPk dPk
+    let transOps = map (uncurry toTransOp) (HM.toList changeset)
+
+    -- Bulid reverse transitive set and convert it to reverseTransOps
+    let reverseTrans = buildReverseTrans changeset
+    reverseOps <- forM (HM.toList reverseTrans) $ \(k, (ad0, dl0)) -> do
+        let ad = HS.toList ad0 -- view patterns are for the weak ones
+            dl = HS.toList dl0
+        prev <- HS.toList <$> GS.getDlgTransitiveReverse k
+        unless (null $ ad `intersect` dl) $ throwM $ DBMalformed $
+            sformat ("Couldn't build reverseOps: ad `intersect` dl is nonzero. "%
+                     "ad: "%listJson%", dl: "%listJson)
+                    ad dl
+        unless (all (`elem` prev) dl) $ throwM $ DBMalformed $
+            sformat ("Couldn't build reverseOps: revtrans has "%listJson%", while "%
+                     "trans says we should delete "%listJson)
+                    prev dl
+        pure $ GS.SetTransitiveDlgRev k $ HS.fromList $ (prev \\ dl) ++ ad
+
+    pure $ SomeBatchOp $ transOps <> reverseOps
+  where
+    {-
+    Get the transitive changeset.
+
+    To scare the reader, we suggest the example. Imagine the following
+    transformation (all arrows are oriented to the right). First graph
+    G is one we store in DB. We apply the list of edgeActions {del DE,
+    add CF} ~ F to it. Note that order of applying edgeActions
+    matters, so we can't go incrementally building graph G' = G ∪ F
+    from G.
+
+    A   F--G      A   F--G
+     \             \ /
+      C--D--E  ⇒    C  D  E
+     /             /
+    B             B
+
+    Let dlg_H(a) denote the transitive delegation relation, returning
+    Nothing or Just dPk. Then we want to:
+    1. Find affected users af = {e ∈ E_G ∪ E_F | dlg_G(e) ≠ dlg_G'(e)}
```

Comment:
I don't get these denotations.
E_G = set of edges of graph G? usually it is denoted as `E(G)`.
And I don't get what does mean `dlg_G(e)` seems `dlg_G :: V -> V`, but `e` is edge